lib/client/connection.js

```diff
@@ -139,10 +139,16 @@ Connection.prototype.bindToSocket = function(socket) {
     }
   };
 
-  socket.onopen = function() {
-    connection._setState('connecting');
+  // If socket is already open, do handshake immediately.
+  // Otherwize do handshake when socket open
+  if (socket.readyState === 1) {
     connection._initializeHandshake();
-  };
+  } else {
+    socket.onopen = function() {
```

Comment:
I'm not sure about hiding this in an `else` statement. For example, I'm pretty sure if you use a reconnecting websocket, you'll still want the `onopen` listener bound so that if your connection drops and gets re-established, this is correctly called.

Reply:
hmmm. I'm not sure if `_initializeHandshake` is idempotent or not?

Reply:
It should eventually just trigger `Connection._initialize` (after a round-trip), which was the behaviour before the handshake was added (previously simply triggered by the `Agent` sending an `init` message on construction, when binding to a new socket).